daemon/config.go

```diff
@@ -72,3 +81,83 @@ func (config *Config) InstallCommonFlags(cmd *flag.FlagSet, usageFn func(string)
 	cmd.StringVar(&config.ClusterStore, []string{"-cluster-store"}, "", usageFn("Set the cluster store"))
 	cmd.Var(opts.NewMapOpts(config.ClusterOpts, nil), []string{"-cluster-store-opt"}, usageFn("Set cluster store options"))
 }
+
+// ReadFromFile loads the configuration from a JSON file.
+func (config *Config) ReadFromFile(configFile string) error {
+	config.reloadLock.Lock()
+	defer config.reloadLock.Unlock()
+
+	f, err := os.Open(configFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return json.NewDecoder(f).Decode(config)
+}
+
+// WatchConfigurationChanges watches for changes in the configuration file
+// and reloads the configuration if this happens.
+func WatchConfigurationChanges(configFile string, reload func(*Config)) {
+	watcher, err := filenotify.New()
+	if err != nil {
+		logrus.Error(err)
+	}
+	defer watcher.Close()
+
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events():
+				logrus.Debug("Event received: %q", event)
+				if event.Op&fsnotify.Write == fsnotify.Write {
+					logrus.Infof("Reloading daemon configuration, %s modified", configFile)
+					newConfig := &Config{}
+					if err := newConfig.ReadFromFile(configFile); err != nil {
+						logrus.Error(err)
+					} else {
+						reload(newConfig)
+					}
+				}
+			case err := <-watcher.Errors():
+				logrus.Error(err)
+			}
+		}
+	}()
+
+	watcher.Add(configFile)
```

Comment:
Note that this can return an error due to things like inotify exhaustion.
This is really where fallback to polling should happen (if you want it to fallback).